Add a `QQBar` interface to GAP

ThomasBreuer opened the pull request:
(addresses issue #4948)

If we decide that the code should be inside the OscarInterface package then this is the occasion to start a package manual for OscarInterface.

ThomasBreuer committed:
add a `QQBar` interface to GAP

fingolfin commented on gap/pkg/OscarInterface/gap/QQBar.gi:
```diff
@@ -0,0 +1,195 @@
+  [ "IsQQBarFieldElement" ],
+  function( x )
+  return QQBarFieldElement( Julia.sqrt( x ) );
+  end );
```
Why does this need an explicit call to `JuliaPointer` in the one case but not in the other?

ThomasBreuer replied:
In both cases, `JuliaPointer` is not needed.
In calls to Julia functions, arguments that are GAP objects with `JuliaPointer` attribute automatically get replaced by their `JuliaPointer` values.

fingolfin reviewed (approved):
Fantastic, that was quick.

fingolfin commented on gap/pkg/OscarInterface/gap/QQBar.gi:
```diff
@@ -0,0 +1,195 @@
+BindGlobal( "QQBarField_Julia", Oscar.algebraic_closure( Oscar.QQ ) );
+
+# We cannot assign the Julia type at load time.
+QQBarFieldElementMatrixType:= fail;
```
Not so important but out of curiosity, why doesn't this work:
```suggestion
QQBarFieldElementMatrixType:= Oscar_jl.eval(Julia.Meta.parse(Julia.String("Matrix{QQBarFieldElem}")));
```

To make this nicer, perhaps we should add a variant of `JuliaEvalString` which takes two arguments: a module and a string.

ThomasBreuer replied:
Yes, this variant works.

ThomasBreuer replied:
Some tests fail, this seems to be the same problem that was addressed in #2333, and I would say the easiest solution for the current problem is the same as there: Bind the variable in question inside Oscar's Julia code, not in GAP code.

fingolfin commented on gap/pkg/OscarInterface/gap/QQBar.gi:
```diff
@@ -0,0 +1,195 @@
+  local MM;
+
+  if QQBarFieldElementMatrixType = fail then
+    QQBarFieldElementMatrixType:= JuliaEvalString( "Matrix{QQBarFieldElem}" );
```
Thinking some more, this won't work if the user did not do `using Oscar` (so e.g. if Oscar is only loaded indirectly)...

ThomasBreuer committed:
address comments

- assign `QQBarFieldElementMatrixType` at load time
- omit unnecessary `JuliaPointer` calls

ThomasBreuer committed:
address test failures

- Do not use `Oscar_jl.eval`, it will result in error messages:
  ERROR: LoadError: InitError: Error thrown by GAP: Error, Evaluation into the closed module `Oscar` breaks incremental compilation because the side effects will not be permanent. This is likely due to some other module mutating `Oscar` with `eval` during precompilation - don't do this.
- Instead, bind the constant in question in Oscar's Julia code.
  (For that, rename `gap/pkg/OscarInterface/julia/alnuth.jl` to `gap/pkg/OscarInterface/julia/constants.jl`,
  and rename the Julia module from `Alnuth` to `OscarInterfaceConstants`.
